airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/state/object_storage/ObjectStorageDestinationStateManager.kt

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.state.object_storage
+
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.databind.node.ObjectNode
+import io.airbyte.cdk.load.command.DestinationStream
+import io.airbyte.cdk.load.file.object_storage.ObjectStorageClient
+import io.airbyte.cdk.load.file.object_storage.PathFactory
+import io.airbyte.cdk.load.state.DestinationState
+import io.airbyte.cdk.load.state.DestinationStatePersister
+import io.airbyte.cdk.load.state.DestinationStatePersisterFactory
+import io.airbyte.cdk.load.util.serializeToString
+import io.airbyte.cdk.util.Jsons
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.annotation.Secondary
+import jakarta.inject.Singleton
+
+class ObjectStorageDestinationState @JsonCreator constructor(generations: List<Generation>) :
+    DestinationState {
+    @JsonIgnore
+    private val generationMap: MutableMap<Long, MutableList<ObjectAndPart>> = mutableMapOf()
+
+    data class Generation(
+        @JsonProperty("generation_id") val generationId: Long,
+        @JsonProperty("objects") val objects: List<ObjectAndPart>
+    )
+
+    data class ObjectAndPart(
+        @JsonProperty("key") val key: String,
+        @JsonProperty("part_number") val partNumber: Long
+    )
+
+    @get:JsonProperty("generations")
+    val generations: List<Generation>
+        get() = generationMap.map { Generation(it.key, it.value) }
+
+    init {
+        generations.forEach { generationMap[it.generationId] = it.objects.toMutableList() }
```

Comment:
why not just store the map directly? afaict your pr currently generates `{generations: [{generation_id: 42, objects: [{key: foo_0.txt, part_number: 0}]}]}`

but storing generationMap would be `{generationMap: {42: [{key: foo_0.txt, part_number: 0}]}}`, which doesn't seem particularly different?

(... is the problem that json keys have to be strings >.> )

Reply:
Yeah, I just assumed the optimal runtime representation wouldn't serialize well. But I didn't actually try to confirm that

Reply:
Works fine, I'll simplify it 😊

Reply:
New rep: ```"""{"generations_by_state":{"FINALIZED":{"0":{"key1":0,"key2":1},"1":{"key3":0,"key4":1}}}}"""```